Sangho internship blog post (#497)
Quansight/Quansight-website: apps/labs/posts/sangho-internship-blogpost.md

---

> title: 'Sangho's Internship at Quansight'

Reviewer: ```suggestion
title: "Sangho's Internship at Quansight"
```

Reviewer: Since you are using a single quote in your title, I suggest changing to double quotes to wrap the title. YAML is good with both quotes like python and I suspect this is why you get the error on the next line.

Author: Can I add my mentor Victor as a reviewr?

> title: "Sangho's Internship at Quansight"

Reviewer: I suggest picking a title which is more related to your project's work, not everyone who reads Quansight Labs' blog will know who you are after all 😉. If I scroll through our Blog pages, how would I know what your internship/blog is about?

published: October 6, 2022
author: sangho-lee
description: 'Blogpost of working on the PyTorch-Ignite project during internship at Quansight'
category: [Machine Learning, OSS Experience]
featuredImage:
  src: /posts/sangho-internship-blogpost/sangho1.png

>   alt: 'An image of the assigned mentors.'

Reviewer: This alt text needs to be updated. The image is of you, not your mentors 🙃

hero:

>   imageSrc: /posts/sangho-internship-blogpost/ignite_logo_mixed.svg

Reviewer: The logo is being cropped at the top. Maybe we can upload a picture with more white space so that doesn't happen?

Author: I'm not sure because I do not find other images. Let me try with `png` format and if it happens again, please let me know :)

  imageAlt: 'PyTorch-Ignite Logo'
---

Hey, I'm [Sangho Lee], a master's student from Seoul National University.
I have participated to the [PyTorch-Ignite](https://github.com/pytorch/ignite) especially with distibuted computations at Quansight Labs.
First thing I contributed was imprving test code of each metrics computated on DDP (Distributed Data Parallel) config.
Second is modifying all_reduce and all_gather methods to take group arguments

## About PyTorch-Ignite and distributed computations

PyTorch-Ignite is a high-level library to help with training and evaluating neural networks in PyTorch flexibly and transparently.

When you have a large batch of dataset such as training images, or videos then your device can’t afford more than a few datasets then it is natural to use multi-gpu in distributed configuration.

In this case, data is transferred to multiple devices for computations then after calculation, it gathers outputs from each devices.

> <img alt="ddp configuration" src="/posts/sangho-blog-post/ddp1.png" />

Reviewer: Our blog rendering likes it when you close html tags explicitly, i.e. with `/img>` rather than `/>`.
You might also consider using markdown for images if you aren't using any explicit html to place the image in a certain way on teh page. E.g. ![ddp configuration](/posts/sangho-blog-post/ddp1.png).

---
title: "Sangho's Internship at Quansight with PyTorch-Ignite project"

> published: October 6, 2022

Reviewer: New release date

author: sangho-lee
description: 'Blogpost of working on the PyTorch-Ignite project during internship at Quansight'
category: [Machine Learning, OSS Experience]
featuredImage:
  src: /posts/sangho-internship-blogpost/sangho1.png
  alt: 'An image of the intern, Sangho.'
hero:
  imageSrc: /posts/sangho-internship-blogpost/ignite_logo_mixed.png
  imageAlt: 'PyTorch-Ignite Logo'
---

Hey, I'm [Sangho Lee], a master's student from Seoul National University.

> I have participated in the [PyTorch-Ignite](https://pytorch-ignite.ai/) project internship at Quansight Labs working on test code improvements and features for distributed computations.

Reviewer: ```suggestion
I have participated in the [PyTorch-Ignite](https://pytorch-ignite.ai/) project internship at Quansight Labs, working on test code improvements and features for distributed computations.
```

The first part of my contributions is improvements to the test code for metric computation in Distributed Data Parallel (DDP) configuration.
Then I worked on adding the `group` argument to the `all_reduce` and `all_gather` methods in [`ignite.distributed`](https://pytorch.org/ignite/distributed.html) module.

## [About PyTorch-Ignite and distributed computations](https://pytorch-ignite.ai/tutorials/advanced/01-collective-communication/)

> PyTorch-Ignite is a high-level library to help with training and evaluating neural networks in PyTorch flexibly and transparently.

Reviewer: ```suggestion
PyTorch-Ignite is a high-level library which helps with training and evaluating neural networks in PyTorch flexibly and transparently.
```

to simplify distributed configuration setup for multiple backends.

## How I contributed with improving test code in DDP config

Problem : test code for metrics computation has incorrectly implemented correctness checks in a distributed configuration

There are 3 items to be checked to ensure that the test code for each metric works correctly in DDP configuration.
1) Generate random input data on each rank and make sure it is different on each rank. This input data is used to compute a metric value with ignite.
2) Gather data with [`idist.all_gather`](https://pytorch-ignite.ai/tutorials/advanced/01-collective-communication/#all-gather) on each rank such that they have the same data before computing reference metric
3) Compare the computed metric value with the reference value (e.g. computed with scikit-learn)

## How I contributed new feature: `group` argument to `all_reduce` and `all_gather` methods

Problem : Existing methods in PyTorch-Ignite uses all ranks, however, for certain use cases users may want to choose a subset of ranks for collecting the data like in the picture.

> As mentioned, Distributed part of Ignite is a wrapper of different backends like [horovod](https://horovod.ai/), [nccl](https://developer.nvidia.com/nccl), [gloo](https://github.com/facebookincubator/gloo) and [xla](https://github.com/pytorch/xla).

Reviewer: ```suggestion
As mentioned, the distributed part of Ignite is a wrapper of different backends like [horovod](https://horovod.ai/), [nccl](https://developer.nvidia.com/nccl), [gloo](https://github.com/facebookincubator/gloo) and [xla](https://github.com/pytorch/xla).
```

I added a new group method for generating group depending on its backend and modified all_reduce and all_gather to take group arguments for users to select the devices.
![Code snippets](/posts/sangho-internship-blogpost/code1.png)

### My contributions

![Contributions with improving test](/posts/sangho-internship-blogpost/cont1.png)
![Contributions with improving test](/posts/sangho-internship-blogpost/cont2.png)

### What I learned

> Last 3 months was really precious time for me as an intern of Quansight.

Reviewer: ```suggestion
These 3 months were really precious time for me as an intern of Quansight.
```

---
title: "Sangho's Internship at Quansight with PyTorch-Ignite project"
published: October 6, 2022
author: sangho-lee
description: 'Blogpost of working on the PyTorch-Ignite project during internship at Quansight'
category: [Machine Learning, OSS Experience]
featuredImage:
  src: /posts/sangho-internship-blogpost/sangho1.png
  alt: 'An image of the intern, Sangho.'
hero:
  imageSrc: /posts/sangho-internship-blogpost/ignite_logo_mixed.png
  imageAlt: 'PyTorch-Ignite Logo'
---

Hey, I'm [Sangho Lee], a master's student from Seoul National University.

> I have participated to the [PyTorch-Ignite](https://pytorch-ignite.ai/) project especially with distributed computations at Quansight Labs.

Reviewer 2: ```suggestion
I have participated in the [PyTorch-Ignite](https://pytorch-ignite.ai/) project internship at Quansight Labs working on test code improvements and features for distributed computations.
```

> First thing I contributed was improvements of the test code for metric computation in Distributed Data Parallel (DDP) configuration.

Reviewer 2: ```suggestion
First part of my contributions is improvements of the test code for metric computation in Distributed Data Parallel (DDP) configuration.
```

> Second is modifying `all_reduce` and `all_gather` methods to take `group` argument.

Reviewer 2: ```suggestion
Then I worked on adding `group` argument to `all_reduce` and `all_gather` methods in [`ignite.distributed`](https://pytorch.org/ignite/distributed.html) module.
```

## [About PyTorch-Ignite and distributed computations](https://pytorch-ignite.ai/tutorials/advanced/01-collective-communication/)

> When you have 1) a larger batch of dataset such as training images or you 2) need to parallelize computations, then it is natural to use multi-gpu in distributed configuration.

Reviewer 2: @puhuk Can you rewrite this section using https://pytorch.org/tutorials/beginner/dist_overview.html as source info and maybe others from internet.

PyTorch-Ignite is a high-level library to help with training and evaluating neural networks in PyTorch flexibly and transparently.
PyTorch-Ignite distributed module (ignite.distributed) is a helper module to use distributed settings for multiple backends like `nccl`, `gloo`, `mpi`, `xla` and `horovod`.

![idist configuration](/posts/sangho-internship-blogpost/ddp0.png)

By simply designating the current backend, ignite.distributed provides a context manager to simplify the code of distributed configuration setup for all above supported backends. And it also provides methods for using existing configuration from each backend.
For example, `auto_model`, `auto_optim` or `auto_dataloader` helps provided model to adapt existing configuration and `Parallel` helps 
to simplify distributed configuration setup for multiple backends.

When using distributed computation with PyTorch-Ignite, data is transferred to multiple devices for computations then after calculation, it gathers outputs from each devices.
![ddp configuration](/posts/sangho-internship-blogpost/ddp1.png)

> ## How PyTorch gathers data

Reviewer 2: ```suggestion
## Collective functions in PyTorch
```

There are two major ways of collecting outputs from each devices, all reduce and all gather
All reduce : Make each device have same values of summation
![All reduce](/posts/sangho-internship-blogpost/allreduce.png)
All gather : Make each device have same values of list

> ![All gather](/posts/sangho-internship-blogpost/allgather.png)

Reviewer 2: Do we really need this section ?

## How I contributed with improving test code in DDP config

> Problems : Many test code already implemented did not test correctly for distributed data pararall environments

Reviewer 2: ```suggestion
Problem : test code for metrics computation has incorrectly implemented correctness checks in distributed configuration
```

> There are 3 items to be checked to ensure that the test codes for each metric work correctly in DDP configuration.

Reviewer 2: ```suggestion
There are 3 items to be checked to ensure that the test code for each metric works correctly in DDP configuration.
```

> 1) Generate random data differently from each device to check metric from gathered data has different value with metric calculated from each rank.

Reviewer 2: ```suggestion
1) Generate random input data on each rank and make sure it is different on each rank. This input data is used to compute metric value with ignite.
```

> 2) Gather data with [`idist.all_gather`](https://pytorch-ignite.ai/tutorials/advanced/01-collective-communication/#all-gather) or [`idist.all_reduce`](https://pytorch-ignite.ai/tutorials/advanced/01-collective-communication/#all-reduce) to each device have same data before computing metric

Reviewer 2: ```suggestion
2) Gather data with [`idist.all_gather`](https://pytorch-ignite.ai/tutorials/advanced/01-collective-communication/#all-gather) on each rank such that they have same data before computing reference metric
```

> 3) Compare calculated metric with reference result from scikit-learn

Reviewer 2: ```suggestion
3) Compare computed metric value with the reference value (e.g. computed with scikit-learn)
```

> ## How I contributed with all_reduce and all_gather to take group argument

Reviewer 2: ```suggestion
## How I contributed new feature: `group` argument to `all_reduce` and `all_gather` methods
```

> Problems : Existing methods in PyTorch-Ignite uses all ranks, however, for certain use-cases users may want to choose a subset of ranks for collecting the data like in the picture.

Reviewer 2: ```suggestion
Problem : Existing methods in PyTorch-Ignite uses all ranks, however, for certain use-cases users may want to choose a subset of ranks for collecting the data like in the picture.
```

Distributed part of Ignite is wrapper of different backend like [horovod](https://horovod.ai/), [nccl](https://developer.nvidia.com/nccl), [gloo](https://github.com/facebookincubator/gloo) and [xla](https://github.com/pytorch/xla).
I added new group method for generating group depends on its backend and modified all_reduce and all_gather to take group arguments for users to select the devices.
![Code snippets](/posts/sangho-internship-blogpost/code1.png)

### My contributions

![Contributions with improving test](/posts/sangho-internship-blogpost/cont1.png)
![Contributions with improving test](/posts/sangho-internship-blogpost/cont2.png)

### What I learned

Last 3 months was really precious time for me as an intern of Quansight.

> PS: I want to thank my mentors [Victor Fomin](https://github.com/vfdev-5) for the teaching and support during the internship.

Reviewer 2: ```suggestion
PS: I want to thank my mentor [Victor Fomin](https://github.com/vfdev-5) for the teaching and support during the internship.
```
